File: go/gate_test.go
package koosgate

import (
	"testing"
	"time"
)

var now = time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)

func lease(subject string, expiresIn time.Duration) AuthorityLease {
	return AuthorityLease{
		Subject:     subject,
		Scope:       "deploy",
		Purpose:     "deploy-service",
		Constraints: map[string]string{"risk": "low"},
		IssuedAt:    now.Add(-time.Hour),
		ExpiresAt:   now.Add(expiresIn),
		Issuer:      "human-chair",
	}
}

func evidence(subject string, validity bool, status EpistemicStatus, expiresIn time.Duration) EvidenceResult {
	return EvidenceResult{
		Subject:         subject,
		Claims:          map[string]string{"integrity": "ok"},
		Source:          "adapter",
		ObservedAt:      now.Add(-time.Minute),
		IssuedAt:        now.Add(-time.Minute),
		ExpiresAt:       now.Add(expiresIn),
		Validity:        validity,
		Provenance:      "adapter:tpm",
		EpistemicStatus: status,
	}
}

func actor(authority []AuthorityLease) Actor {
	return Actor{ID: "wolf-1", Capabilities: map[string]bool{"deploy": true}, Authority: authority}
}

func request() GateRequest {
	return GateRequest{Subject: "wolf-1", Capability: "deploy", Scope: "deploy", PolicySatisfied: true, ConflictMode: Deny}
}

func TestGateContractMatrix(t *testing.T) {
	gate := Gate{}
	tests := []struct {
		name     string
		actor    Actor
		request  GateRequest
		evidence []EvidenceResult
		assert   func(*testing.T, GateDecision)
	}{
		{"test 01", actor(nil), request(), []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny || decision.Predicates["A"] {
				t.Fatalf("expected deny on missing authority: %+v", decision)
			}
		}},
		{"test 02", actor(nil), request(), []EvidenceResult{{Subject: "wolf-1", Claims: map[string]string{"pi_verified": "true"}, Source: "verifier", ObservedAt: now, IssuedAt: now, ExpiresAt: now.Add(time.Hour), Validity: true, Provenance: "math:proof", EpistemicStatus: Verified}}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny || !decision.Predicates["E"] || decision.Predicates["A"] {
				t.Fatalf("expected deny with valid evidence but no authority: %+v", decision)
			}
		}},
		{"test 03", actor(nil), request(), []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny {
				t.Fatalf("expected deny: %+v", decision)
			}
		}},
		{"test 04", actor(nil), request(), []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny {
				t.Fatalf("expected deny: %+v", decision)
			}
		}},
		{"test 05", actor([]AuthorityLease{lease("wolf-1", time.Hour)}), request(), nil, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny || decision.Predicates["E"] {
				t.Fatalf("expected deny on missing evidence: %+v", decision)
			}
		}},
		{"test 06", actor([]AuthorityLease{lease("wolf-1", time.Hour)}), request(), []EvidenceResult{evidence("other", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny || decision.Predicates["S"] {
				t.Fatalf("expected deny on subject mismatch: %+v", decision)
			}
		}},
		{"test 07", actor([]AuthorityLease{lease("wolf-1", -time.Hour)}), request(), []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Deny || decision.Predicates["A"] || decision.Predicates["T"] {
				t.Fatalf("expected deny on expired lease: %+v", decision)
			}
		}},
		{"test 08", actor([]AuthorityLease{lease("wolf-1", time.Hour)}), GateRequest{Subject: "wolf-1", Capability: "deploy", Scope: "deploy", RequiredConstraints: map[string]string{"risk": "low"}, PolicySatisfied: true, ConflictMode: Deny}, []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != Allow {
				t.Fatalf("expected allow: %+v", decision)
			}
		}},
		{"test 10", actor([]AuthorityLease{lease("wolf-1", time.Hour)}), GateRequest{Subject: "wolf-1", Capability: "deploy", Scope: "deploy", PolicySatisfied: true, ConflictMode: HumanReview}, []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour), evidence("wolf-1", false, Contradicted, time.Hour)}, func(t *testing.T, decision GateDecision) {
			if decision.Decision != HumanReview || !decision.EvidenceConflict {
				t.Fatalf("expected human review: %+v", decision)
			}
		}},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			decision := gate.Evaluate(tc.actor, tc.request, tc.evidence, now)
			tc.assert(t, decision)
		})
	}
}

func TestDecisionRuntimeSeparationAndLedger(t *testing.T) {
	gate := Gate{}
	runtime := Runtime{}
	decision := gate.Evaluate(actor([]AuthorityLease{lease("wolf-1", time.Hour)}), GateRequest{Subject: "wolf-1", Capability: "deploy", Scope: "deploy", RequiredConstraints: map[string]string{"risk": "low"}, PolicySatisfied: true, ConflictMode: Deny}, []EvidenceResult{evidence("wolf-1", true, Verified, time.Hour)}, now)
	if decision.Decision != Allow {
		t.Fatalf("expected allow: %+v", decision)
	}
	event := runtime.Execute(decision, "deploy-service", false)
	if event.Status != Failed {
		t.Fatalf("expected runtime failure after allow: %+v", event)
	}
	ledger := &Ledger{}
	ledger.Append("decision", map[string]string{"decision": string(decision.Decision)}, now)
	ledger.Append("event", map[string]string{"status": string(event.Status)}, now)
	if len(ledger.Entries) != 2 || ledger.Entries[0].EntryType != "decision" || ledger.Entries[1].EntryType != "event" {
		t.Fatalf("expected decision and event history: %+v", ledger.Entries)
	}
}

func TestPackCharacters(t *testing.T) {
	if len(PackCharacters) != 13 || PackCharacters[0].Name != "QUILL" || PackCharacters[12].Name != "NYX" {
		t.Fatalf("expected full pack roster: %+v", PackCharacters)
	}
}

File: typescript/gate.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';

import { gate, runtime, Ledger, ObjectEvidenceAdapter, PACK_CHARACTERS, type Actor, type AuthorityLease, type EvidenceResult } from './gate.ts';

const now = new Date('2026-01-01T00:00:00Z');

function lease(subject = 'wolf-1', expiresHours = 1): AuthorityLease {
  return {
    subject,
    scope: 'deploy',
    purpose: 'deploy-service',
    constraints: { risk: 'low' },
    issued_at: new Date('2025-12-31T23:00:00Z'),
    expires_at: new Date(now.getTime() + expiresHours * 60 * 60 * 1000),
    issuer: 'human-chair',
  };
}

function evidence(subject = 'wolf-1', validity = true, status: EvidenceResult['epistemic_status'] = 'VERIFIED', expiresHours = 1): EvidenceResult {
  return {
    subject,
    claims: { integrity: 'ok' },
    source: 'adapter',
    observed_at: new Date('2025-12-31T23:55:00Z'),
    issued_at: new Date('2025-12-31T23:55:00Z'),
    expires_at: new Date(now.getTime() + expiresHours * 60 * 60 * 1000),
    validity,
    provenance: 'adapter:tpm',
    epistemic_status: status,
  };
}

function actor(authority: AuthorityLease[] = []): Actor {
  return { id: 'wolf-1', capabilities: ['deploy'], authority };
}

test('01 missing authority denies', () => {
  const decision = gate(actor(), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [evidence()], now);
  assert.equal(decision.decision, 'DENY');
  assert.equal(decision.predicates.A, false);
});

test('02 perfect evidence without authority denies', () => {
  const adapter = new ObjectEvidenceAdapter('verifier');
  const perfect = adapter.toEvidence({
    subject: 'wolf-1',
    claims: { pi_verified: true },
    observed_at: new Date('2025-12-31T23:59:00Z'),
    issued_at: new Date('2025-12-31T23:59:00Z'),
    expires_at: new Date('2026-01-01T01:00:00Z'),
    validity: true,
    provenance: 'math:proof',
    epistemic_status: 'VERIFIED',
  });
  const decision = gate(actor(), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [perfect], now);
  assert.equal(decision.decision, 'DENY');
  assert.equal(decision.predicates.E, true);
  assert.equal(decision.predicates.A, false);
});

test('03 valid TPM quote without authority denies', () => {
  assert.equal(gate(actor(), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [evidence()], now).decision, 'DENY');
});

test('04 capability without authority denies', () => {
  assert.equal(gate(actor(), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [evidence()], now).decision, 'DENY');
});

test('05 authority without evidence denies', () => {
  const decision = gate(actor([lease()]), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [], now);
  assert.equal(decision.decision, 'DENY');
  assert.equal(decision.predicates.E, false);
});

test('06 subject mismatch denies', () => {
  const decision = gate(actor([lease()]), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [evidence('other')], now);
  assert.equal(decision.decision, 'DENY');
  assert.equal(decision.predicates.S, false);
});

test('07 expired lease denies', () => {
  const decision = gate(actor([lease('wolf-1', -1)]), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy' }, [evidence()], now);
  assert.equal(decision.decision, 'DENY');
  assert.equal(decision.predicates.A, false);
  assert.equal(decision.predicates.T, false);
});

test('08 all predicates satisfied allows', () => {
  const decision = gate(actor([lease()]), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy', required_constraints: { risk: 'low' } }, [evidence()], now);
  assert.equal(decision.decision, 'ALLOW');
  assert.ok(Object.values(decision.predicates).every(Boolean));
});

test('09 gate allow does not imply runtime success', () => {
  const decision = gate(actor([lease()]), { subject: 'wolf-1', capability: 'deploy', scope: 'deploy', required_constraints: { risk: 'low' } }, [evidence()], now);
  const event = runtime(decision, 'deploy-service', false);
  const ledger = new Ledger();
  ledger.append({ entry_type: 'decision', recorded_at: now, payload: { decision: decision.decision } });
  ledger.append({ entry_type: 'event', recorded_at: now, payload: { status: event.status } });
  assert.equal(decision.decision, 'ALLOW');
  assert.equal(event.status, 'FAILED');
  assert.deepEqual(ledger.entries().map((entry) => entry.entry_type), ['decision', 'event']);
});

test('10 conflicting evidence can escalate to human review', () => {
  const decision = gate(
    actor([lease()]),
    { subject: 'wolf-1', capability: 'deploy', scope: 'deploy', conflict_mode: 'HUMAN_REVIEW' },
    [evidence('wolf-1', true, 'VERIFIED'), evidence('wolf-1', false, 'CONTRADICTED')],
    now,
  );
  assert.equal(decision.decision, 'HUMAN_REVIEW');
  assert.equal(decision.evidence_conflict, true);
});

test('pack roster is complete', () => {
  assert.equal(PACK_CHARACTERS.length, 13);
  assert.equal(PACK_CHARACTERS[0].name, 'QUILL');
  assert.equal(PACK_CHARACTERS.at(-1)?.name, 'NYX');
});
